#################
## libraries ####
#################
options(import.path=c("~/analysis/fractalgenetics",
                      "~/projects"))
options(bitmapType = 'cairo', device = 'pdf')


modules::import_package('ggplot2', attach=TRUE)
modules::import_package('GGally', attach=TRUE)
optparse <- modules::import_package('optparse')
ukbtools <- modules::import_package('ukbtools')
related <- modules::import('GWAS/relatedness')
autofd <- modules::import('fractal-analysis-processing')
smooth <- modules::import('utils/smoothAddR2')
plinkqc <- modules::import_package('plinkQC')


makeCall <- function(genotype) {
    x <- rep(1, length(genotype))
    x[genotype < 0.5] <- 0
    x[genotype >= 1.5] <- 2
    return(x)
}

#################################
## parameters and input data ####
#################################
option_list <- list(
    make_option(c("-u", "--ukbdir"), action="store", dest="rawdir",
               type="character", help="Path to ukb directory with decrypted ukb
               key.html file [default: %default].", default=NULL),
    make_option(c("-o", "--outdir"), action="store", dest="outdir",
               type="character", help="Path to output directory [default:
               %default].", default=NULL),
    make_option(c("-p", "--pheno"), action="store", dest="pheno",
               type="character", help="Path to fd phenotype file [default:
               %default].", default=NULL),
    make_option(c("-c", "--cov"), action="store", dest="cov",
               type="character", help="Path to LV volume covariate file
               [default: %default].", default=NULL),
    make_option(c("-i", "--interpolate"), action="store", dest="interpolate",
               type="integer", help="Number of slices to interpolate to
               [default: %default].", default=9),
    make_option(c("-s", "--samples"), action="store", dest="samples",
               type="character", help="Path to ukb genotype samples file
               [default: %default].", default=NULL),
    make_option(c("-r", "--relatedness"), action="store", dest="relatedness",
               type="character", help="Path to relatedness file generated by
               ukbgene rel [default: %default].", default=NULL),
    make_option(c("-e", "--europeans"), action="store", dest="europeans",
               type="character", help="Path to European samples file generated
               by ancestry.smk [default: %default].", default=NULL),
    make_option(c("-pcs", "--pcs"), action="store", dest="pcs",
               type="character", help="Path to pca output file generated by
               flashpca [default: %default].", default=NULL),
    make_option(c("--debug"), action="store_true",
               dest="debug", default=FALSE, type="logical",
               help="If set, predefined arguments are used to test the script
               [default: %default].")
)

args <- optparse$parse_args(OptionParser(option_list=option_list))

if (args$debug) {
    args <- list()
    args$outdir <- "~/data/digital-heart/phenotype/FD"
    args$pheno <- "~/data/digital-heart/phenotype/FD/20191002_DCM_FD_all.csv"
    args$interpolate <- 9
    args$cov <- "~/data/digital-heart/phenotype/2Dphenotype/20160412_All_BRU_format.txt"
    args$samples <- "~/data/digital-heart/genotype/imputation/combined/genotypes/gencall.combined.clean.related.chr1.sample"
    args$europeans <- "~/data/digital-heart/genotype/QC/combined/DCM.gencall.combined.clean.related.fam"
    args$path2plink <- "/homes/hannah/bin/plink"
    args$genodir <- "/homes/hannah/data/digital-heart/genotype/QC/combined"
}

#############
## data  ####
#############

# genotypes
genotypes_all <- data.table::fread("~/data/digital-heart/gwas/DCM/DCM.FD.Pseudomultitrait_Slices_sig5e08.sigSNPs.dosage",
                               data.table=FALSE, stringsAsFactors=FALSE)
genotypes <- genotypes_all
rownames(genotypes) <- genotypes$rsid
genotypes <- genotypes[, -c(1:6)]
genotypes <- t(genotypes)

## Filter European HVOLs for related samples
# European HVOLs
european_dcm <- data.table::fread(args$europeans, data.table=FALSE,
                            stringsAsFactors=FALSE)


# QC on DCMs
missing <- plinkqc$check_het_and_miss(indir=args$genodir,
                                     name="DCM.gencall.combined.clean.related",
                                     interactive=FALSE, verbose=TRUE,
                                     path2plink=args$path2plink)
related <- plinkqc$check_relatedness(indir=args$genodir,
                                     name="DCM.gencall.combined.clean.related",
                                     interactive=FALSE, verbose=TRUE,
                                     path2plink=args$path2plink)

missingness <- data.table::fread(file.path(args$genodir,
                                 "DCM.gencall.combined.clean.related.imiss"),
                                 data.table=FALSE,
                                 stringsAsFactors=FALSE)

relatedness <- data.table::fread(file.path(args$genodir,
                                 "DCM.gencall.combined.clean.related.genome"),
                                 data.table=FALSE,
                                 stringsAsFactors=FALSE)



## FD measurements ####
dataFD <- data.table::fread(args$pheno, data.table=FALSE,
                            stringsAsFactors=FALSE, na.strings=c("NA", "NaN"))

rownames(dataFD) <- dataFD[, 1]
colnames(dataFD)[grepl('FD.*Slice.1',colnames(dataFD))] <- 'Slice 1'
dataFD <- dataFD[,grepl("Slice.\\d{1,2}", colnames(dataFD))]
colnames(dataFD) <- gsub("Slice.(.*)", "Slice\\1", colnames(dataFD))

# Exclude individuals where less than 6 slices were measured
NaN_values <- c("Sparse myocardium", "Meagre blood pool","FD measure failed",
                "No mask present")
fd_notNA <- apply(dataFD, 1,  function(x) {
                length(which(!(is.na(x) | x %in% NaN_values))) > 5
            })
dataFD <- dataFD[fd_notNA, ]

## 2D phenotypes ####
covariates <- read.table(args$cov, stringsAsFactors=FALSE,  fill=TRUE, sep="\t",
                                      quote = "", header=TRUE)

# all digital-heart imputed genotypes in bgen format
samples <- data.table::fread(args$samples, data.table=FALSE, skip=2,
                             stringsAsFactors=FALSE)[,1:2]
colnames(samples) <- c("FID", "IID")

################
## analysis ####
################

# Exclude related and non-european individuals
relatedness <- relatedness[relatedness$IID1 %in% rownames(dataFD) |
                           relatedness$IID2 %in% rownames(dataFD),]

fail_related <- plinkqc$relatednessFilter(relatedness=relatedness,
                          otherCriterion=missingness,
                          otherCriterionMeasure = "F_MISS",
                          relatednessTh=0.185,
                          otherCriterionTh = 0.1,
                          otherCriterionThDirection = "gt")

relatedIDs <- as.character(fail_related$failIDs$IID)
dcm <- european_dcm[!european_dcm[,1] %in% relatedIDs,]
dataFD <- dataFD[rownames(dataFD) %in% dcm[,1],]

# covariates
colnames(covariates) <- c("IID", "ethnicity", "sex", "diagnosis",  "age",
                          "height", "weight")
covariates$sex[covariates$sex ==""] <- NA
covariates$sex <- as.numeric(as.factor(covariates$sex))
covariates$bmi <- covariates$weight/(covariates$height/100)^2
covariates$bsa <- sqrt(covariates$weight * covariates$height/3600)
rownames(covariates) <- covariates$IID
covariates <- covariates[!apply(covariates, 1, function(x) any(x==0)),]
covariates <- dplyr::select(covariates, sex, age, weight, height, bmi)
covariates <- covariates[rownames(covariates) %in% rownames(dataFD),]

# interpolate FD slice measures
FDi <- autofd$interpolate$fracDecimate(data=dataFD,
                                       interpNoSlices=args$interpolate,
                                       id.col.name='rownames')
# summary fd measurements
summaryFDi <- data.frame(t(apply(as.matrix(FDi), 1,
                                          autofd$stats$summaryStatistics,
                       discard=FALSE, sections="BMA")))

# combine everything
fd_all <- merge(summaryFDi[,-c(1,2,4,6,8)], FDi, by=0)
fd_all <- merge(fd_all, covariates, by.x=1, by.y=0)
fd_all$sex <- as.factor(fd_all$sex)


# plot distribution of FD along heart
FDalongHeart <- reshape2::melt(FDi, value.name = "FD")
colnames(FDalongHeart)[1:2] <- c("ID", "Slice")

FDalongHeart$Slice <- as.factor(as.numeric(gsub("Slice_", "",
                                                FDalongHeart$Slice)))
FDalongHeart$Location <- "Apical section"
FDalongHeart$Location[as.numeric(FDalongHeart$Slice) <= 3] <- "Basal section"
FDalongHeart$Location[as.numeric(FDalongHeart$Slice) <= 6 &
                      as.numeric(FDalongHeart$Slice) > 3] <- "Mid section"
FDalongHeart$Location <- factor(FDalongHeart$Location,
                                levels=c("Basal section", "Mid section",
                                         "Apical section"))

p_fd <- ggplot(data=FDalongHeart)
p_fd <- p_fd + geom_boxplot(aes(x=Slice, y=FD, color=Location)) +
    scale_color_manual(values=c('#67a9cf','#1c9099','#016c59')) +
    labs(x="Slice", y="FD") +
    theme_bw()

ggsave(plot=p_fd, file=paste(args$outdir, "/FDAlongHeart_DCM_slices",
                             args$interpolate, ".pdf", sep=""),
       height=4, width=4, units="in")



## Plot distribution of covariates ####
df <- dplyr::select(fd_all, MeanBasalFD, MeanMidFD, MeanApicalFD,
                    sex, age, height, weight, bmi)
p <- ggpairs(df,
             upper = list(continuous = wrap("density", col="#b30000",
                                            size=0.1)),
             diag = list(continuous = wrap("densityDiag", size=0.4)),
             lower = list(continuous = wrap("smooth", alpha=0.5,size=0.1,
                                            pch=20),
                          combo = wrap("facethist")),
             columnLabels = c("meanBasalFD", "meanMidFD",
                              "meanApicalFD",
                              "Sex~(f/m)", "Age~(years)", "Height~(m)",
                              "Weight~(kg)", "BMI~(kg/m^2)"),
             labeller = 'label_parsed',
             axisLabels = "show") +
    theme_bw() +
    theme(panel.grid = element_blank(),
          axis.text = element_text(size=6),
          axis.text.x = element_text(angle=90),
          strip.text = element_text(size=8),
          strip.background = element_rect(fill="white", colour=NA))
ggsave(plot=p, file=paste(args$outdir, "/pairs_fdcovariates_DCM.png", sep=""),
       height=12, width=12, units="in")

# Regress covariates
fd_reg <- sapply(2:13, function(x) {
    cat(x)
    df <- data.frame(trait=fd_all[,x], fd_all[,14:18])
    lm(trait ~ ., data=df)$residuals
})
colnames(fd_reg) <- colnames(fd_all)[2:13]
rownames(fd_reg) <- fd_all$Row.names

# merge FD measures and genotypes
fd_geno <- merge(fd_reg, genotypes, by=0)
write.table(fd_geno, file=paste(args$outdir, "/DCM_FD_genotypes.csv", sep=""),
            sep=',', col.names=NA, row.names=TRUE, quote=FALSE)

lead_snps <- c("rs6587924",
               "rs35770803",
               "rs1892027",
               "rs71394376",
               "rs4677294",
               "rs1918978",
               "rs10076436",
               "rs3130976",
               "rs9320648",
               "rs6981461",
               "rs35006907",
               "rs7132327",
               "rs71105784",
               "rs17608766",
               "rs113394178",
               "rs3788488",
               "rs11570470",
               "rs117154502",
               "rs78033733")
fd_geno_lead <- merge(fd_reg, genotypes[,colnames(genotypes) %in% lead_snps],
                      by=0)
write.table(fd_geno_lead, file=paste(args$outdir,
                                     "/DCM_FD_genotypes_lead_snps.csv", sep=""),
            sep=',', col.names=NA, row.names=TRUE, quote=FALSE)

fd_geno_lead <- read.table(file=paste(args$outdir,
                                      "/DCM_FD_genotypes_lead_snps.csv", sep=""),
                           sep=',', header=TRUE)
snp_cols <- grepl("^rs", colnames(fd_geno_lead))
fd_geno_lead_hard <- fd_geno_lead
fd_geno_lead_hard[, snp_cols] <- apply(fd_geno_lead_hard[, snp_cols], 2,
                                       makeCall)
fd_geno_lead_hard <- fd_geno_lead_hard[,-1]
fd_geno_lead_hard$Row.names <- as.character(fd_geno_lead_hard$Row.names)
colnames(fd_geno_lead_hard)[1] <- "IID"

##
ukbgeno_all <-  data.table::fread("~/Dropbox/FD paper/Pseudomultitrait_slices_sig5e08_genotypes.dosage.gz", data.table=FALSE)
ukbgeno <- ukbgeno_all
ukbgeno <- dplyr::filter(ukbgeno, rsid %in% lead_snps)
rownames(ukbgeno) <- ukbgeno$rsid
ukbgeno <- ukbgeno[, -c(1:6)]
ukbgeno <- t(ukbgeno)
ukbgeno_hard <- apply(ukbgeno,2,makeCall)
rownames(ukbgeno_hard) <- rownames(ukbgeno)

ukbslices <-  data.table::fread("~/data/ukbb/ukb-hrt/phenotypes/180628_fractal_dimension/FD_slices_EUnorel.csv", data.table=FALSE)
ukbsummary <-  data.table::fread("~/data/ukbb/ukb-hrt/phenotypes/180628_fractal_dimension/FD_summary_EUnorel.csv", data.table=FALSE)
ukbpheno <- merge(ukbsummary, ukbslices, by=1)
ukbcovs <-  data.table::fread("~/data/ukbb/ukb-hrt/phenotypes/180628_fractal_dimension/FD_covariates_EUnorel.csv", data.table=FALSE)
ukbreg <- apply(ukbpheno[,-1],2, function(x) {
    df <- data.frame(trait=x, ukbcovs)
    lm(trait ~ ., data=df)$residuals
})
rownames(ukbreg) <- ukbpheno[,1]
ukball <- merge(ukbreg, ukbgeno_hard, by=0)
colnames(ukball)[1] <- "IID"


## combine datasets
common_dh <- fd_geno_lead_hard[,colnames(fd_geno_lead_hard) %in% colnames(ukball)]
common_ukb <- ukball[,colnames(ukball) %in% colnames(fd_geno_lead_hard)]

common_dh$study <- "DCM"
common_ukb$study <- "UKB"
combined <- rbind(common_dh, common_ukb)
saveRDS(combined, "~/data/digital-heart/gwas/FD/ukb_DCM_FD_regressed.rds")


## reformat data
combined_snps <- pivot_longer(combined, cols=starts_with("rs"),
                              values_to="genotype", names_to="rsID")
combined_slices <- pivot_longer(combined, cols=starts_with("Slice") ,
                                values_to="FD", names_to="Pheno") %>%
    select(-starts_with("Mean"))
combined_regions <-  pivot_longer(combined, cols=starts_with("Mean") ,
                                  values_to="FD", names_to="Pheno") %>%
    select(-starts_with("Slice"))
combined_slices_snps <- pivot_longer(combined_snps, cols=starts_with("Slice") ,
                                     values_to="FD", names_to="Pheno") %>%
    select(-starts_with("Mean"))
combined_regions_snps <- pivot_longer(combined_snps, cols=starts_with("Mean") ,
                                      values_to="FD", names_to="Pheno") %>%
    select(-starts_with("Slice"))




# plot combined distribution of FD along heart
combined_slices$Slice <- as.numeric(gsub("Slice_", "", combined_slices$Pheno))
combined_slices$Location <- "Apical section"
combined_slices$Location[combined_slices$Slice<= 3] <- "Basal section"
combined_slices$Location[combined_slices$Slice <= 6 &
                             combined_slices$Slice > 3] <- "Mid section"
combined_slices$Location <- factor(combined_slices$Location,
                                levels=c("Basal section", "Mid section",
                                         "Apical section"))

p_fd <- ggplot(data=combined_slices)
p_fd + geom_boxplot(aes(x=Pheno, y=FD, color=Location, fill=study)) +
    scale_color_manual(values=c('#67a9cf','#1c9099','#016c59')) +
    labs(x="Slice", y="FD") +
    theme_bw()




## test associations
common_snps <-  colnames(select(slices, starts_with("rs")))

## joint pheno and all SNP effects
lmm_slices <- lme(as.formula(paste("FD ~ ",
                                 paste(c(common_snps, "study"), collapse="+"))),
                random = ~ 1|Pheno,
                data=combined_slices, control = lmeControl(opt = "optim"))

lmm_regions <- lme(as.formula(paste("FD ~ ",
                                 paste(c(common_snps, "study"), collapse="+"))),
                random = ~ 1|Pheno,
                data=combined_regions, control = lmeControl(opt = "optim"))


## single pheno and all SNP effects
test <- filter(combined_slices,  Pheno == "Slice_4")
lm_slice <- lm(as.formula(paste("FD ~ ",
                               paste(c(common_snps, "study"), collapse="+"))),
                data=test)
summary(lm_both)$coefficients

test <- filter(combined_regions,  Pheno == "MeanMidFD")
lm_region <- lm(as.formula(paste("FD ~ ",
                                paste(c(common_snps, "study"), collapse="+"))),
               data=test)
summary(lm_region)$coefficients

## joint pheno and single snp effects
sapply(common_snps, function(x) {
    lmm_slices <- lme(as.formula(paste("FD ~ ",
                                   paste(c(x, "study"), collapse="+"))),
                  random = ~ 1|Pheno,
                  data=combined_slices, control = lmeControl(opt = "optim"))
    summary(lmm_slices)$tTable[3,5]
})

sapply(common_snps, function(x) {
lmm_regions <- lme(as.formula(paste("FD ~ ",
                                    paste(c(common_snps, "study"), collapse="+"))),
                   random = ~ 1|Pheno,
                   data=combined_regions, control = lmeControl(opt = "optim"))
    summary(lmm_slices)$tTable[3,5]
})

## Visualise models
test <- filter(combined_slices_snps, rsID  == "rs17608766",  Pheno == "Slice_4")
p <- ggplot(test)
p + geom_boxplot(aes(y=FD, x= as.factor(genotype),fill=study)) +
    scale_fill_brewer(type="qual", palette = 2) +
    ylab("FD on slice 4") +
    xlab("Genotype") +
    theme_bw()

p <- ggplot(test)
p + geom_boxplot(aes(y=FD, x= as.factor(study),fill=study)) +
    scale_fill_brewer(type="qual", palette = 2) +
    facet_wrap(~as.factor(genotype)) +
    ylab("FD all slices") +
    xlab("Genotype") +
    theme_bw()

p <- ggplot(combined_slices_snps,
            aes(x=as.factor(genotype), y=FD, color=study)) +
    geom_violin(aes(y=FD, x= as.factor(genotype)),
                position = position_dodge(width = 0.9)) +
    scale_color_brewer(type="qual", palette = 2) +
    stat_summary(fun.y=median, geom="point", aes(color=study),
                 position = position_dodge(width = 0.9),
                 shape=20) +
    stat_summary(fun.data=mean_sdl, geom="pointrange", aes(color=study),
                 position = position_dodge(width = 0.9),shape=20) +
    ylab("All FD") +
    xlab("Genotype") +
    theme_bw()
 print(p)
